nengo_spa/ast/symbolic.py

```diff
@@ -0,0 +1,148 @@
+"""AST classes for symbolic operations."""
+
+import nengo
+import numpy as np
+
+from nengo_spa.ast.base import infer_types, Fixed
+from nengo_spa.exceptions import SpaTypeError
+from nengo_spa.pointer import SemanticPointer
+from nengo_spa.types import TAnyVocab, TScalar, TVocabulary
+from nengo.utils.compat import is_number
+
+
+def as_symbolic_node(obj):
+    if is_number(obj):
+        obj = FixedScalar(obj)
+    return obj
+
+
+class Symbol(Fixed):
+    @property
+    def expr(self):
+        raise NotImplementedError()
+
+
+class FixedScalar(Symbol):
+    # Does not implement any operators as Python does so for numbers and
+    # once a Python number gets converted into an AST node it must have been
+    # used with some other non-scalar AST node that will return
+    # a non-FixedScalar AST node. Thus, there should never be a situation where
+    # an operator is applied to two FixedScalar nodes.
+
+    def __init__(self, value):
+        super(FixedScalar, self).__init__(type_=TScalar)
+        self.value = value
+
+    def connect_to(self, sink, **kwargs):
+        return nengo.Connection(self.construct(), sink, **kwargs)
+
+    def construct(self):
+        return nengo.Node(self.value, label=str(self.value))
+
+    def evaluate(self):
+        return self.value
+
+    @property
+    def expr(self):
+        return repr(self.value)
+
+
+class PointerSymbol(Symbol):
+    def __init__(self, expr, type_=TAnyVocab):
+        super(PointerSymbol, self).__init__(type_=type_)
+        self._expr = expr
+
+    def connect_to(self, sink, **kwargs):
+        return nengo.Connection(self.construct(), sink, **kwargs)
+
+    def construct(self):
+        return nengo.Node(self.evaluate().v, label=self.expr)
+
+    def evaluate(self):
+        if not isinstance(self.type, TVocabulary):
+            raise SpaTypeError(
+                "Cannot evaluate a symbolic semantic pointer expression "
+                "without knowing the vocabulary.")
+        return self.type.vocab.parse(self.expr)
+
+    @property
+    def expr(self):
+        return self._expr
+
+    def __invert__(self):
+        return PointerSymbol('~' + self.expr, self.type)
+
+    def __neg__(self):
+        return PointerSymbol('-' + self.expr, self.type)
+
+    def __add__(self, other):
+        other = as_symbolic_node(other)
+        if not isinstance(other, PointerSymbol):
+            return NotImplemented
+        type_ = infer_types(self, other)
+        return PointerSymbol(self.expr + '+' + other.expr, type_)
+
+    def __radd__(self, other):
+        return self + other
+
+    def __sub__(self, other):
+        other = as_symbolic_node(other)
```

Comment:
The code block:
```
other = as_symbolic_node(other)
if not isinstance(other, SOMETHING):
     return NotImplemented
type_ = infer_types(self, other)
```

Seems to be repeated a lot, both in `symbolic.py` and in `pointer.py`. Would it be worth refactoring into a function?

Reply:
I added a commit to do such a refactoring as much as possible and reasonable. There are two things that prevent a further reduction of the duplicated code:

1. The type `SOMETHING` is sometimes the class that is just being defined. Thus, we cannot define a decorator because it needed to be instantiated with a reference to that class when it is not yet fully defined. We also cannot use a normal function because in the case that the type does not match the outer function must return `NotImplemented`.
2. The last line `type_ = ...` could be refactored out, but then the value of `type_` needs to be passed in as a function argument which changes the function signature of the operators to a non-default signature. That seems confusing to me and the code duplication seem to be the lesser evil.